### ISSUE TITLE
As with NameGenerator, separate out other code concepts too

### ISSUE THREAD

**Issue by NathanJAdams** [evolvio/evolv.io#28]
The work done to separate out NameGenerator is great and would be very worthwhile to do the same with other code concepts. The files are very large and have several mutual dependencies which makes it difficult for lots of people to make contributions at the same time. It would make contributing much easier if all the code concepts were separated out from each other and the mutual dependencies removed.

Some possibilities are eg.
Environment - Holds temperature information, growth rates and board size
Brain - Holds Axons and Neurons - done
FileSaver - Code to save state - in progress
UserController - Code to respond to mouse and keyboard actions
UI - Code to control and draw the user interface

This is all up for debate, if people have better ideas how to separate code concepts out please say

**Comment by SirSKillz** [comment]
I think this is really important so tasks could be more easily assigned. As it stands now, if more than one person is editing, most likely other people can't help. Furthermore, this would allow newcomers to more easily grasp what each section of the code is used for and how it fits in with the project as a whole. Honestly, I think this is probably the most important thing we could do right now.

**Comment by kodomazer** [comment]
I'm going to make a list of code concepts that we should extract after #70 is completed (hopefully via #79 )

**Comment by kodomazer** [comment]
#83 got me thinking.

When we pull out the actions as a concept, should we also look at pulling out the brain inputs as a concept and have that bridge the gap between creatures and their brains.

e.g.  We have `CreatureInput` and `CreatureAction` interfaces which creatures will implement and pass to the brain, and then the brain will have some internal method of assigning them to various positions in the brain so when `CreatureInput` is called for an input that is where it is put, or if `CreatureAction` is called then that is where it gets its input from to do the action.

While we're at it we can make vision its own object that the creature will instantiate and attach to itself and then pass corresponding input and action classes to the brain so we can get some vision evolution (and of course each eye consumes additional energy, as well as brain neurons)

**Comment by PRabahy** [comment]
@kodomazer Yep, that is pretty much my idea. I'm not really a fan of arrays, so the fact that the brain is implemented as a 3d array right now really bugs me. I think it should the brain should be a list of input actions, a list of neurons, and a list of outputs. In one of my next PRs, I'm going to at least get rid of all the magic constants in the brain/neurons.

**Comment by 10carfernandez** [comment]
I'm writing code for saving files and most of the methods are written in the Board class, but I'll make a FileSaver class for all of those methods now that I'm on it. 

I was also thinking about maybe making a class for the UI controls as well because I found myself having to change several numerical values across different files just to modify some of the buttons.

**Comment by NateUniAccount** [comment]
Closing as covered by roadmap
